engine/evarutil.ml

```diff
@@ -684,15 +685,40 @@ let undefined_evars_of_named_context evd nc =
     nc
     ~init:Evar.Set.empty
 
-let undefined_evars_of_evar_info evd evi =
+(* compute undefined evars, cache results *) 
+let cache_undefined_evars_of_evar_info ev evd evi =
+  let deps =
   Evar.Set.union (undefined_evars_of_term evd (EConstr.of_constr evi.evar_concl))
     (Evar.Set.union
        (match evi.evar_body with
 	 | Evar_empty -> Evar.Set.empty
-	 | Evar_defined b -> undefined_evars_of_term evd (EConstr.of_constr b))
+	 | Evar_defined b ->
+	    (* we assume ev is undefined, so body should not be defined *)
+	    assert false)
```

Comment:
You should rather used undefined_evar_of_term for the body here, and remove the special casing in contained_in_info below